### PR TITLE
[Phase 3] Sequencer Component Templates

### PR DESCRIPTION
**Pull request by tamingchaos**
## Summary

Implements complete Helm templates for sequencer component with Service, ConfigMap, and persistence.

### Service (RD-532)
- Headless Service for StatefulSet
- Exposes NATS port 4222 when embedded NATS enabled
- Standard label selectors for pod discovery

### ConfigMap (RD-534)
- NATS server configuration with JetStream settings
- Port configuration (4222)
- Storage limits (1GB memory, 10GB file)
- Only created when embedded NATS is enabled

### Persistence (RD-535)
- Data volumeClaimTemplate for blockchain data
  - Configurable size (default 100Gi)
  - Configurable storageClass
  - Mounted at `/data`
- NATS storage volumeClaimTemplate for JetStream
  - Configurable size (default 10Gi)
  - Configurable storageClass
  - Mounted at `/data/nats`
- ConfigMap volume mount at `/etc/nats`

## Testing

**TDD Workflow:** All templates developed using RED → GREEN → REFACTOR

- ✅ 6 service tests passing
- ✅ 5 configmap tests passing
- ✅ 6 persistence tests passing
- ✅ 7 statefulset tests passing (existing + updated)
- ✅ **24 total tests passing**
- ✅ helm lint clean

```bash
# Run tests
helm unittest k8s/helm

# Lint
helm lint k8s/helm
```

## Next Steps

- RD-536: Integration test for sequencer with embedded NATS in K8s
- RPC component templates (RD-533, RD-537, RD-538)

## Linear Tasks

Completes: RD-532, RD-534, RD-535

### CODE REVIEW

File: k8s/helm/templates/sequencer/statefulset.yaml [hunk 1]
```diff
@@ -33,6 +39,49 @@ spec:
...
+        {{- end }}
+        {{- if and .Values.sequencer.nats.embedded .Values.sequencer.nats.persistence.enabled }}
+        - name: nats-storage
+          mountPath: /data/nats
+        {{- end }}
+        {{- if .Values.sequencer.nats.embedded }}
+        - name: nats-config
+          mountPath: /etc/nats
+          readOnly: true
+        {{- end }}
```
[the remark this answers is not in the document]
Comment (tamingchaos) [reply k8s/helm/templates/sequencer/statefulset.yaml:72]: Already addressed - line 42 wraps volumeMounts with `{{- if or .Values.sequencer.persistence.enabled .Values.sequencer.nats.embedded }}` so it only renders when needed.

File: k8s/helm/templates/sequencer/configmap.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    server_name: {{ include "cdk-erigon.fullname" . }}-sequencer
+
+    jetstream {
+      store_dir: "/data/nats"
+      max_memory_store: 1073741824
+      max_file_store: 10737418240
+    }
+
+    max_payload: 8388608
+    max_pending: 67108864
```
[the remark this answers is not in the document]
Comment (tamingchaos) [reply k8s/helm/templates/sequencer/configmap.yaml]: Already addressed - values are templated from `.Values.sequencer.nats.maxMemoryStore`, `.maxFileStore`, `.maxPayload`, `.maxPending` (lines 16-21 in configmap.yaml, configurable in values.yaml lines 37-43).

File: k8s/helm/values.yaml [hunk 3]
```diff
@@ -32,6 +32,16 @@ sequencer:
     # -- Enable embedded NATS server
     embedded: true
 
+    # -- NATS JetStream configuration
```
[the remark this answers is not in the document]
Comment (tamingchaos) [reply k8s/helm/values.yaml]: Already addressed - line 35 now uses `# NATS JetStream configuration` without the `--` prefix.

File: k8s/helm/templates/sequencer/configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+{{- if .Values.sequencer.enabled }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ include "cdk-erigon.fullname" . }}-sequencer
+  labels:
+    {{- include "cdk-erigon.labels" . | nindent 4 }}
+    component: sequencer
+data:
+  {{- if .Values.sequencer.nats.embedded }}
```
[the remark this answers is not in the document]
Comment (tamingchaos) [reply k8s/helm/templates/sequencer/configmap.yaml]: Already addressed - outer condition is now `{{- if and .Values.sequencer.enabled .Values.sequencer.nats.embedded }}` so ConfigMap only renders when NATS is enabled.

File: k8s/helm/templates/sequencer/configmap.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+    jetstream {
+      store_dir: "/data/nats"
+      max_memory_store: {{ printf "%.0f" .Values.sequencer.nats.maxMemoryStore }}
+      max_file_store: {{ printf "%.0f" .Values.sequencer.nats.maxFileStore }}
+    }
+
+    max_payload: {{ printf "%.0f" .Values.sequencer.nats.maxPayload }}
+    max_pending: {{ printf "%.0f" .Values.sequencer.nats.maxPending }}
+  {{- end }}
```
[the remark this answers is not in the document]
Comment (tamingchaos) [reply k8s/helm/templates/sequencer/configmap.yaml]: Already addressed in commit 5f1836e - removed the redundant inner conditional.

File: k8s/helm/templates/sequencer/statefulset.yaml [hunk 6]
```diff
@@ -23,6 +23,12 @@ spec:
         runAsNonRoot: {{ .Values.sequencer.securityContext.runAsNonRoot }}
         runAsUser: {{ .Values.sequencer.securityContext.runAsUser }}
         fsGroup: {{ .Values.sequencer.securityContext.fsGroup }}
```
[the remark this answers is not in the document]
Comment (tamingchaos) [reply k8s/helm/templates/sequencer/statefulset.yaml:26]: Fixed - added `runAsGroup` and `seccompProfile` to the security context.

File: k8s/helm/templates/sequencer/statefulset.yaml [hunk 7]
```diff
@@ -33,6 +39,51 @@ spec:
...
+        {{- if and .Values.sequencer.nats.embedded .Values.sequencer.nats.persistence.enabled }}
+        - name: nats-storage
+          mountPath: /data/nats
+        {{- end }}
+        {{- if .Values.sequencer.nats.embedded }}
+        - name: nats-config
+          mountPath: /etc/nats
+          readOnly: true
+        {{- end }}
+        {{- end }}
```
[the remark this answers is not in the document]
Comment (tamingchaos) [reply k8s/helm/templates/sequencer/statefulset.yaml]: This was already fixed in a previous commit - volumeMounts is now unconditionally present with individual items conditionally rendered. The nats-config mount renders when `sequencer.nats.embedded` is true (lines 65-68).

File: k8s/helm/tests/sequencer_pvc_test.yaml [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+          content:
+            name: nats-config
+            configMap:
+              name: RELEASE-NAME-cdk-erigon-sequencer
+      - contains:
+          path: spec.template.spec.containers[0].volumeMounts
+          content:
+            name: nats-config
+            mountPath: /etc/nats
+            readOnly: true
```
[the remark this answers is not in the document]
Comment (tamingchaos) [reply k8s/helm/tests/sequencer_pvc_test.yaml:91]: Good point - will add edge case test for NATS enabled with both persistence options disabled.

File: k8s/helm/tests/sequencer_configmap_test.yaml [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+      - exists:
+          path: data["nats.conf"]
+
+  - it: should not contain NATS config when embedded NATS is disabled
+    set:
+      sequencer.enabled: true
+      sequencer.nats.embedded: false
+    asserts:
+      - notExists:
+          path: data["nats.conf"]
```
[the remark this answers is not in the document]
Comment (tamingchaos) [reply k8s/helm/tests/sequencer_configmap_test.yaml]: Acknowledged - this test is indeed redundant since the ConfigMap does not render when embedded NATS is disabled. Will remove or update.
